doc/services/pm/device.rst

```diff
@@ -20,90 +27,175 @@ in power saving mode. This method allows saving power even when the CPU is
-One of the macro parameters is the pointer to the PM action callback.
-If the driver doesn't implement any power control operations, it can initialize
-the corresponding pointer with ``NULL``.
+Drivers initialize devices using macros. See :ref:`device_model_api` for
+details on how these macros are used. A driver which implements device power
+management support must provide these macros with arguments that describe its
+power management implementation.
+
+Use :c:macro:`PM_DEVICE_DEFINE` or :c:macro:`PM_DEVICE_DT_DEFINE` to define the
+power management resources required by a driver. These macros allocate the
+driver-specific state which is required by the power management subsystem.
+
+Drivers can use :c:macro:`PM_DEVICE_GET` or
+:c:macro:`PM_DEVICE_DT_GET` to get a pointer to this state. These
+pointers should be passed to ``DEVICE_DEFINE`` or ``DEVICE_DT_DEFINE``
+to initialize the power management field in each ``struct device``.
+
+Here is some example code showing how to implement device power management
+support in a device driver.
+
+.. code-block:: c
+
+    #define DT_DRV_COMPAT dummy_device
+
+    static int dummy_driver_pm_action(const struct device *dev,
+                                      enum pm_device_action *action)
+    {
+        switch (action) {
+        case PM_DEVICE_ACTION_SUSPEND:
+            /* suspend the device */
+            ...
+            break;
+        case PM_DEVICE_ACTION_RESUME:
+            /* resume the device */
+            ...
+            break;
+        case PM_DEVICE_ACTION_TURN_ON:
+            /*
+	     * powered on the device, used when the power
+	     * domain this device belongs is resumed.
+	     */
+            ...
+            break;
+        case PM_DEVICE_ACTION_TURN_OFF:
+            /*
+	     * power off the device, used when the power
+	     * domain this device belongs is suspended.
+	     */
+            ...
+            break;
+        default:
+            return -ENOTSUP;
+        }
+
+        return 0;
+    }
+
+    PM_DEVICE_DT_INST_DEFINE(0, dummy_driver_pm_action);
+
+    DEVICE_DT_INST_DEFINE(0, &dummy_init,
+        PM_DEVICE_DT_INST_GET(0), NULL, NULL, POST_KERNEL,
+        CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
 
 .. _pm-device-busy:
 
 Busy Status Indication
 **********************
 
-The SOC interface executes some power policies that can turn off power to devices,
-causing them to lose their state. If the devices are in the middle of some
-hardware transaction, like writing to flash memory when the power is turned
-off, then such transactions would be left in an inconsistent state. This
-infrastructure guards such transactions by indicating to the SOC interface that
-the device is in the middle of a hardware transaction.
+When the system is idle and the SoC is going to sleep, the power management
+subsystem can suspend devices, as described in :ref:`pm-device-system-pm`. This
+can cause device hardware to lose some states. Suspending a device which is in
```

Comment:
Looks like a typo sneaked in here in the latest push. It should be "lose some state".

Reply:
weird, "ambiguous" was fixed, your comment even has an "outdated" label. Don't know what revision you have looked.
The other indeed is there, fixing it.